lib/surface/compiler.ex

```diff
@@ -343,6 +346,62 @@ defmodule Surface.Compiler do
      }}
   end
 
+  defp convert_node_to_ast(
+         :for_else,
+         {_name, attributes, children, node_meta},
+         compile_meta
+       ) do
+    meta = Helpers.to_meta(node_meta, compile_meta)
+    default = %AST.AttributeExpr{value: false, original: "", meta: meta}
+    generator = attribute_value_as_ast(attributes, "each", :generator, default, compile_meta)
+
+    [for_children, else_children] =
+      case children do
+        [{:default, [], default, _}, {"#else", _, _, _} = else_block] ->
+          [default, [else_block]]
+
+        [{:default, [], default, _}] ->
+          [default, []]
+
+        children ->
+          [children, []]
+      end
+
+    for_ast = %AST.For{
+      generator: generator,
+      children: to_ast(for_children, compile_meta),
+      meta: meta
+    }
+
+    if else_children == [] do
+      {:ok, for_ast}
+    else
+      [{_, _, [{var, _, _}, _]} | _] = generator.value
+      generator_expr = generator.value ++ [[do: {var, [], nil}]]
+
+      condition = %AST.AttributeExpr{
+        original: "",
+        value:
+          quote do
+            unquote({:for, [generated: true], generator_expr})
+            |> case do
+              [] -> false
+              _ -> true
+            end
+          end,
+        meta: compile_meta
+      }
```

Comment:
> The condition of the If must test against the result of the for. We need to generate something like:
> 
> ```elixir
> for {i <- items, i in [...]} do
>   ...
> end |> case do
>   [] ->
>     # The else block
>   result ->
>     result
> end
> ```

@msaraiva new attempt, a bit closer to what you suggest and handles the generator with filters.

Still not too happy with the code, though.